### PR TITLE
Improve time and memory of plot_initial_state

### LINKED ISSUE #663: plot_initial_state.py is taking forever and running out of memory
Even for meshes that aren't particularly large by our standards (e.g. EC30to60), `plot_initial_state.py` is taking a very long time to run and is running out of memory:
```
Traceback (most recent call last):
  File "./run.py", line 40, in <module>
    subprocess.check_call(['./plot_initial_state.py'])
  File "/usr/projects/climate/SHARED_CLIMATE/anaconda_envs/base/envs/test_compass_0.1.9/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['./plot_initial_state.py']' died with <Signals.SIGKILL: 9>.
Traceback (most recent call last):
  File "./run.py", line 77, in <module>
  File "/usr/projects/climate/SHARED_CLIMATE/anaconda_envs/base/envs/test_compass_0.1.9/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['./run.py']' returned non-zero exit status 1.
slurmstepd: error: Detected 1 oom-kill event(s) in step 17109912.batch cgroup. Some of your processes may have been killed by the cgroup out-of-memory handler.
```

### PR DESCRIPTION
**Pull request by xylar**
This merge converts plot_initial_state to use `xarray` with chunking to use `dask` parallelism.  It also uses `xarray`'s version of
histograms.  The plotting of the largest COMPASS mesh (`ARM`) goes for timing out or running out of memory even for 3 hour jobs to running in 15 seconds and using about 1% of the memory of a Grizzly compute node.

closes #663

### CODE REVIEW

**Comment by xylar** [comment]
## Testing

So far, I tested only on `ARM` on a Grizzly compute node and only on an already existing `initial_state.nc`.  I' have not tested on other test cases, login nodes, or performing the full COMPASS workflow yet.

File: testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py [hunk 1]
```diff
@@ -29,10 +29,11 @@ def main():
     args = parser.parse_args()
 
     # load mesh variables
-    ncfile = Dataset(args.input_file_name, 'r')
-    nCells = ncfile.dimensions['nCells'].size
-    nEdges = ncfile.dimensions['nEdges'].size
-    nVertLevels = ncfile.dimensions['nVertLevels'].size
+    chunks = {'nCells': 32768, 'nEdges': 32768}
+    ds = xarray.open_dataset(args.input_file_name, chunks=chunks)
```
Comment (xylar) [inline testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py:33]: This breaks the data set into "chunks" of 32k cells or edges per "chunk".  Chunks are distributed across multiple threads (36 on a Grizzly compute node) whenever a computation like `min()`, `max()`, `xarray.plot.hist()`, etc. is performed.

File: testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py [hunk 2]
```diff
@@ -51,63 +52,70 @@ def main():
 
     plt.subplot(3, 3, 2)
     varName = 'maxLevelCell'
-    var = ncfile.variables[varName]
-    maxLevelCell = var[:]
-    plt.hist(var, bins=nVertLevels - 4)
+    var = ds[varName]
+    maxLevelCell = var.values - 1
```
Comment (xylar) [inline testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py:56]: I always define the index variables right away to be their zero-based python variants.

File: testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py [hunk 3]
```diff
@@ -51,63 +52,70 @@ def main():
...
+    xarray.plot.hist(var, bins=nVertLevels - 4)
```
Comment (xylar) [inline testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py:57]: This is the key: the `xarray` version of `hist` is **way** better, it seems.

File: testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py [hunk 4]
```diff
@@ -51,63 +52,70 @@ def main():
...
     plt.ylabel('frequency')
     plt.xlabel(varName)
-    txt = '{}{:9.2e} {:9.2e} {}\n'.format(txt, np.amin(var), np.amax(var), varName)
+    txt = '{}{:9.2e} {:9.2e} {}\n'.format(txt, var.min().values, 
+                                          var.max().values, varName)
```
Comment (xylar) [inline testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py:61]: The `xarray` version of `min` and `max` know to ignore masked (`NaN`) entries.

File: testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py [hunk 5]
```diff
@@ -51,63 +52,70 @@ def main():
...
-        cell1 = cellsOnEdge[:, 1]-1
+        cellMask[:, k] = k <= maxLevelCell
+        cell0 = cellsOnEdge[:, 0]
+        cell1 = cellsOnEdge[:, 1]
         edgeMask[:, k] = np.logical_and(np.logical_and(cellMask[cell0, k],
                                                        cellMask[cell1, k]),
                                         np.logical_and(cell0 >= 0,
                                                        cell1 >= 0))
+    cellMask = xarray.DataArray(data=cellMask, dims=('nCells', 'nVertLevels'))
+    edgeMask = xarray.DataArray(data=edgeMask, dims=('nEdges', 'nVertLevels'))
```
Comment (xylar) [inline testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py:83]: By making these masks into `xarray.DataArray`s, we can use them to mask other data arrays with `where()`.

File: testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py [hunk 6]
```diff
@@ -51,63 +52,70 @@ def main():
...
 
     plt.subplot(3, 3, 4)
     varName = 'temperature'
-    var = ncfile.variables[varName][0, :, :][cellMask]
-    plt.hist(var, bins=100, log=True)
+    var = ds[varName].isel(Time=0).where(cellMask)
```
Comment (xylar) [inline testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py:87]: Select time index 0 of the given variable, then mask it by putting `NaN`s for invalid cells.

File: testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py [hunk 7]
```diff
@@ -51,63 +52,70 @@ def main():
...
+    txt = '{}{:9.2e} {:9.2e} {}\n'.format(txt, var.min().values, 
+                                          var.max().values, varName)
 
-    rx1Edge = ncfile.variables['rx1Edge']
     plt.subplot(3, 3, 7)
     varName = 'rx1Edge'
-    var = ncfile.variables[varName][0, :, :][edgeMask]
-    plt.hist(var, bins=100, log=True)
+    var = ds[varName].isel(Time=0).where(edgeMask)
+    maxRx1Edge = var.max().values
```
Comment (xylar) [inline testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py:113]: compute the max Haney number after masking.

File: testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py [hunk 8]
```diff
@@ -117,7 +125,9 @@ def main():
     plt.text(0, 1, txt, verticalalignment='top', fontproperties=font)
     plt.axis('off')
 
-    plt.savefig(args.output_file_name)
+    plt.tight_layout(pad=4.0)
+
+    plt.savefig(args.output_file_name, bbox_inches='tight', pad_inches=0.1)
```
Comment (xylar) [inline testing_and_setup/compass/ocean/scripts/plots/plot_initial_state.py:130]: The layout of the figure is a lot better with these settings, taken from MPAS-Analysis climatology map plots.

**Comment by xylar** [comment]
Here's the plot for ARM:
initial_state

**Comment by mark-petersen** [comment]
Wow, this is fantastic, and serves as a great example of how to convert a post-processing code to xarray and dask. Thank you, @xylar! I'll test this with upcoming batch of COMPASS changes.

**Pull request merged**